File: common/src/main/java/net/impleri/itemskills/utils/ListDiff.java
package net.impleri.itemskills.utils;

import net.impleri.itemskills.ItemHelper;
import net.minecraft.world.item.Item;

import java.util.ArrayList;
import java.util.List;

public class ListDiff {
    public static boolean contains(List<Item> a, List<Item> b) {
        return getMissing(a, b).size() == 0;
    }

    public static List<Item> getMissing(List<Item> a, List<Item> b) {
        if (a.size() == 0 && b.size() == 0) {
            return new ArrayList<>();
        }

        var bStrings = b.stream()
                .map(ItemHelper::getItemKey)
                .toList();

        return a.stream()
                .filter(existing -> !bStrings.contains(ItemHelper.getItemKey(existing)))
                .toList();
    }
}

File: common/src/main/kotlin/net/impleri/itemskills/util/ListDiff.kt
package net.impleri.itemskills.util

import net.impleri.itemskills.api.ItemRestriction
import net.minecraft.world.item.Item

object ListDiff {
  fun contains(a: List<Item>, b: List<Item>): Boolean {
    return getMissing(a, b).isEmpty()
  }

  fun getMissing(a: List<Item>, b: List<Item>): List<Item> {
    if (a.isEmpty() && b.isEmpty()) {
      return ArrayList()
    }

    val bStrings = b.map { ItemRestriction.getName(it) }

    return a.filter { !bStrings.contains(ItemRestriction.getName(it)) }
  }
}
